src/mame/amstrad/pda600_copro.h

```diff
@@ -0,0 +1,65 @@
+// license:BSD-3-Clause
+// copyright-holders:Sandro Ronco
+/**********************************************************************
+
+    Amstrad PDA600 Coprocessor HLE
+
+**********************************************************************/
+
+#ifndef MAME_MACHINE_PDA600_COPRO_H
+#define MAME_MACHINE_PDA600_COPRO_H
+
+#pragma once
+
+#include "diserial.h"
+
+//**************************************************************************
+//  TYPE DEFINITIONS
+//**************************************************************************
+
+// ======================> pda600_copro_device
+
+class pda600_copro_device : public device_t,
+							public device_buffered_serial_interface<1 + 1 + 255 + 1>
+{
+public:
+	pda600_copro_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	auto tx_callback()    { return m_tx_cb.bind(); }
+	auto tone_callback()  { return m_tone_cb.bind(); }
+
+	DECLARE_WRITE_LINE_MEMBER( wakeup_w );
+	DECLARE_WRITE_LINE_MEMBER( reset_w )    { if (state)  device_reset(); }
+	DECLARE_WRITE_LINE_MEMBER( write_txd )  { rx_w(state); }
+
+private:
+	// device-level overrides
+	virtual void device_resolve_objects() override;
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	// device_serial_interface overrides
+	void tra_callback() override { m_tx_cb(transmit_register_get_data_bit()); }
+	void received_byte(u8 byte) override;
+
+	TIMER_CALLBACK_MEMBER(update_timer);
+	void send_byte(u8 byte);
+	void send_resp();
+	void exec_beep();
+	void exec_train();
+	void exec_recognize();
+	u8 recognize_char();
+
+	devcb_write_line m_tx_cb;
+	devcb_write8     m_tone_cb;
+	emu_timer *      m_update_timer;
+	u8               m_state = 0;
+	u8               m_resp_type = 0;
+	u8               m_resp_data = 0;
+	std::vector<u8>  m_buf;
```

Comment:
Save states don’t work with `std::vector` if it's size changes on-the-fly.

Also, please don’t initialise members at declaration in classes with out-of-line constructors.  There‘s too much potential for confusion as the member can be initialised differently in the constructor initialiser list and declaration.